opal/threads/wait_sync.h

```diff
@@ -75,6 +89,9 @@ static inline int sync_wait_st (ompi_wait_sync_t *sync)
  */
 static inline void wait_sync_update(ompi_wait_sync_t *sync, int updates, int status)
 {
+    sync->signaling = true;
+    /* ensure the signalling value is commited before updating the count */
+    opal_atomic_wmb ();
     if( OPAL_LIKELY(OPAL_SUCCESS == status) ) {
         if( 0 != (OPAL_THREAD_ADD32(&sync->count, -updates)) ) {
```

Comment:
Nice catch. The scope of this race condition is restricted to cases where
the synchronization primitive is used in a wait or waitall but not waitany
or waitsome (where the sync primitive is atomically removed from the
requests before release).

As the only thing we need is to guarantee that the signaling thread is out
of the signaling sequence, so we can start with the signaling field set to
true. This will save the memory barrier in the wait_sync_update, and will
only require the change in WAIT_SYNC_SIGNAL.

Also the error path is now broken, the following patch is necessary.
```
diff --git a/opal/threads/wait_sync.h b/opal/threads/wait_sync.h
index b29f01c..d2b5003 100644
--- a/opal/threads/wait_sync.h
+++ b/opal/threads/wait_sync.h
@@ -80,9 +80,10 @@ static inline void wait_sync_update(ompi_wait_sync_t
*sync, int updates, int sta
             return;
         }
     } else {
-        /* this is an error path so just use the atomic */
-        opal_atomic_swap_32 (&sync->count, 0);
+        /* We need the status to be visible when the count become 0 */
         sync->status = OPAL_ERROR;
+        opal_atomic_wmb();
+        opal_atomic_swap_32 (&sync->count, 0);
     }
     WAIT_SYNC_SIGNAL(sync);
 }
```

On Sun, Jun 26, 2016 at 3:03 PM, Artem Polyakov <notifications@github.com>
wrote:

> In opal/threads/wait_sync.h
> <https://github.com/open-mpi/ompi/pull/1816#discussion_r68502768>:
>
> > @@ -75,6 +89,9 @@ static inline int sync_wait_st (ompi_wait_sync_t *sync)
> >   */
> >  static inline void wait_sync_update(ompi_wait_sync_t *sync, int updates, int status)
> >  {
> > +    sync->signaling = true;
> > +    /* ensure the signalling value is commited before updating the count */
> > +    opal_atomic_wmb ();
> >      if( OPAL_LIKELY(OPAL_SUCCESS == status) ) {
> >          if( 0 != (OPAL_THREAD_ADD32(&sync->count, -updates)) ) {
>
> I guess I'm not correct here.
> Only one (the last) thread will get 0.
> So no race here.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/open-mpi/ompi/pull/1816/files/d204f6740d9dd867bd233d9373f2eb2252cfb827#r68502768>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAnOjScSitDA8_JRaZOxU-NBKbnu_6Ggks5qPni2gaJpZM4I-HOt>
> .
>